source/simulator/core.cc

```diff
@@ -2393,72 +2393,32 @@ namespace aspect
         else
           statistics.add_value("Time step size (seconds)", time_step);
 
-
         // see if we have to start over with a new refinement cycle
         // at the beginning of the simulation
-        if ((timestep_number == 0) &&
-            (pre_refinement_step < parameters.initial_adaptive_refinement))
+        if (timestep_number == 0)
           {
-            if (parameters.timing_output_frequency ==0)
-              computing_timer.print_summary ();
-
-            output_statistics();
-
-            if (parameters.run_postprocessors_on_initial_refinement)
-              postprocess ();
-
-            refine_mesh (max_refinement_level);
-            ++pre_refinement_step;
-            goto start_time_iteration;
+            if (pre_refinement_step < parameters.initial_adaptive_refinement)
+              {
+                do_initial_refinement_cycle(max_refinement_level);
+                ++pre_refinement_step;
+                goto start_time_iteration;
+              }
+            else
+              // invalidate the value of pre_refinement_step since it will no longer be used from here on
+              pre_refinement_step = std::numeric_limits<unsigned int>::max();
           }
 
-        // invalidate the value of pre_refinement_step since it will no longer be used from here on
-        if ( timestep_number == 0 )
-          pre_refinement_step = std::numeric_limits<unsigned int>::max();
-
         postprocess ();
 
-        // see if this is a time step where additional refinement is requested
-        // if so, then loop over as many times as this is necessary
-        if ((parameters.additional_refinement_times.size() > 0)
-            &&
-            (parameters.additional_refinement_times.front () < time+time_step))
-          {
-            while ((parameters.additional_refinement_times.size() > 0)
-                   &&
-                   (parameters.additional_refinement_times.front () < time+time_step))
-              {
-                ++max_refinement_level;
-                refine_mesh (max_refinement_level);
+        // see if this is a time step where refinement is requested
+        if (time_for_additional_refinement_cycle())
+          do_additional_refinement_cycle(max_refinement_level);
+        else if (time_for_regular_refinement_cycle())
+          refine_mesh (max_refinement_level);
```

Comment:
like for the cases below, everything you have (the 4 lines here + what's in at least `do_additional_refinement_cycle`) here could probably be combined into a function `maybe_refine_mesh()`.